File: crates/router/src/destinations.rs
//! Destinations — the per-dispatch sequence table sagas and process managers
//! stamp emitted commands from. The coordinator supplies one next-sequence
//! per output domain (config-driven); the component is a translator, not a
//! decision maker, so it never rebuilds destination state — it only stamps.
//!
//! Transliterated from client-go's `destinations.go`.

use std::collections::HashMap;

use crate::error::{codes, extras, messages, CodedError};
use crate::pb;

/// Destination next-sequences keyed by domain, as supplied by the
/// coordinator in the dispatch request.
pub struct Destinations {
    sequences: HashMap<String, u32>,
}

impl Destinations {
    /// Wraps the coordinator-supplied sequence map.
    pub fn new(sequences: HashMap<String, u32>) -> Self {
        Destinations { sequences }
    }

    /// The next sequence for a domain, or None when the coordinator
    /// supplied none.
    pub fn sequence_for(&self, domain: &str) -> Option<u32> {
        self.sequences.get(domain).copied()
    }

    /// True when a sequence exists for the domain.
    pub fn has(&self, domain: &str) -> bool {
        self.sequences.contains_key(domain)
    }

    /// The domains carrying a sequence (unordered).
    pub fn domains(&self) -> Vec<String> {
        self.sequences.keys().cloned().collect()
    }

    /// Stamps every page of `cmd` with the next sequence for `domain`.
    /// A domain with no supplied sequence is the coded
    /// MISSING_DESTINATION_SEQUENCE (check output_domains config).
    pub fn stamp_command(&self, cmd: &mut pb::CommandBook, domain: &str) -> Result<(), CodedError> {
        let Some(seq) = self.sequences.get(domain).copied() else {
            return Err(CodedError::invalid_argument(
                codes::MISSING_DESTINATION_SEQUENCE,
                messages::MISSING_DESTINATION_SEQUENCE,
                [(extras::DOMAIN.to_string(), domain.to_string())],
            ));
        };
        for page in &mut cmd.pages {
            page.header = Some(pb::PageHeader {
                sequence_type: Some(pb::page_header::SequenceType::Sequence(seq)),
                ..Default::default()
            });
        }
        Ok(())
    }
}

#[cfg(test)]
#[path = "destinations.test.rs"]
mod destinations_tests;

File: bindings/java/src/main/java/io/angzarr/router/Destinations.java
package io.angzarr.router;

import io.angzarr.CommandBook;
import io.angzarr.CommandPage;
import io.angzarr.PageHeader;
import java.util.List;
import java.util.Map;

/**
 * The coordinator-supplied next-sequences for command stamping. Sagas and
 * process managers are translators — they stamp emitted commands, they do not
 * rebuild destination state to make decisions.
 */
public final class Destinations {
  private final Map<String, Integer> sequences;

  /** Wraps a domain→next-sequence map (null becomes empty). */
  public Destinations(Map<String, Integer> sequences) {
    this.sequences = sequences == null ? Map.of() : sequences;
  }

  /** Returns the next sequence for a domain, or null if none exists. */
  public Integer sequenceFor(String domain) {
    return sequences.get(domain);
  }

  /** Reports whether a sequence exists for the domain. */
  public boolean has(String domain) {
    return sequences.containsKey(domain);
  }

  /** Every domain carrying a sequence (unordered). */
  public List<String> domains() {
    return List.copyOf(sequences.keySet());
  }

  /**
   * Returns a copy of {@code cmd} with every page stamped with the next sequence
   * for {@code domain}. A domain with no supplied sequence is the coded {@code
   * MISSING_DESTINATION_SEQUENCE} (check output_domains config).
   */
  public CommandBook stampCommand(CommandBook cmd, String domain) {
    Integer seq = sequences.get(domain);
    if (seq == null) {
      throw new CodedError(
          "MISSING_DESTINATION_SEQUENCE",
          "no sequence for destination domain",
          GrpcCode.INVALID_ARGUMENT,
          Map.of("domain", domain));
    }
    CommandBook.Builder book = cmd.toBuilder();
    for (CommandPage.Builder page : book.getPagesBuilderList()) {
      page.setHeader(PageHeader.newBuilder().setSequence(seq));
    }
    return book.build();
  }
}
